src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -140,6 +140,12 @@ bool CanConsumeRowNonSequentially
         /// </summary>
         char[]? _tempCharBuf;
 
+        /// <summary>
+        /// Used to keep track of every unique row this reader object ever traverses.
+        /// This is used to detect whether inner DbDataReaders are still valid.
+        /// </summary>
+        internal ulong UniqueRowId;
```

Comment:
Take a look at how manage this problem for GetStream/GetTextReader - NpgsqlDataReader has a _columnStream which gets set when GetStream is called, and disposed when moving to the next row.

I don't think there's a particular advantage in any of the two techniques, but we should unify and have just one mechanism for this lifecycle management.

Reply:
Note that an outer NpgsqlDataReader could be referenced by multiple inner data readers, and multiple levels of inner data readers. I also want this to work as good as possible in case we recycle data readers. So not sure how the stream approach would work in this case.

Reply:
You're right - since we need to handle multiple "dependents", the stream approach doesn't work. But then we can convert the stream to look at the new UniqueRowId mechanism.

Reply:
BTW that's also maybe a thing to consider... NpgsqlDataReader currently doesn't support multiple streams even when in non-sequential mode (where there's no reason not too). IIRC this was intentional - I wanted users to not depend on the ability to have multiple ones, so that if they switch to sequential (which generally makes a lot of sense where column streams are used!) they have no issues.

The same logic may apply for the inner data reader, i.e. we may want to restrict to only one active at a given time, just like stream. I doubt this would be restrictive for actual users. What do you think?

Reply:
The "stream approach" wouldn't work for the multiple depth case I think, since you have multiple inner data readers then. If there is a simple way to support multiple streams, I would say, why restrict it just because we haven't seen a good use case so far? But to be consistent with sequential mode, maybe we should still restrict it... I'm not sure.

Reply:
Yeah, it's a bit trickier... I think we need to make sure there's no more than one nested reader open for a given reader (nested or top-level) - that should provide the right protection for sequential. BTW this protection also applies between streams and readers - you can't have both a stream and a reader open at the same time in sequential.

As to restricting in non-sequential as well... I thought about it again, and I guess it's not a problem if working code starts throwing when you switch on sequential - after all this is already what happens if you're not accessing fields in order; sequential simply does impose further constraints on how you interact with the reader. So I'd be fine with removing all restrictions for non-sequential (including the existing stream one).

Reply:
Ok so can we leave "allow multiple streams to be opened for non-sequential mode on a particular row" for a separate PR?

Reply:
Sure!